### DISCUSSION TITLE
"SIMD must divide IFMChannels" error but SIMD does divide IFMChannels

### DISCUSSION THREAD

**Discussion by julienmas** [Xilinx/finn#666] [Q&A] [answered]
Hi everyone,

I want to deploy a custom CNN on my Ultra96V2 board. I am trying to create a custom build flow, inspired from the end2end example notebooks tfc and cnv. When I call the transformation `model = model.transform(ZynqBuild(platform = test_pynq_board, period_ns = target_clk_ns))`, I get the error `AssertionError: SIMD must divide IFMChannels`, even though I made sure that SIMD values divides the IFMChannels values.
Here is my .onnx dataflow model [unet_dataflow_model.txt](https://github.com/Xilinx/finn/files/9328595/unet_dataflow_model.txt) (I put a .txt instead of .onnx to be able to upload it here) and the values I use for folding :
```
fc_layers = unet.get_nodes_by_op_type("MatrixVectorActivation")
# each tuple is (PE, SIMD, in_fifo_depth) for a layer
folding = [
    (4, 27, 32),
    (4, 72, 32),
    (1, 144, 32),
    (4, 72, 32),
    (2, 72, 32),
    (4, 72, 32),
    (2, 72, 32),
    (4, 72, 32),
    (2, 72, 32),
    (4, 72, 32),
    (1, 1, 32)
]
for fcl, (pe, simd, ififodepth) in zip(fc_layers, folding):
    fcl_inst = getCustomOp(fcl)
    fcl_inst.set_nodeattr("PE", pe)
    fcl_inst.set_nodeattr("SIMD", simd)
    fcl_inst.set_nodeattr("inFIFODepth", ififodepth)

# use same SIMD values for the sliding window operators
swg_layers = unet.get_nodes_by_op_type("ConvolutionInputGenerator")
for i in range(len(swg_layers)):
    swg_inst = getCustomOp(swg_layers[i])
    simd = folding[i][1]
    swg_inst.set_nodeattr("SIMD", simd)
```

I tried to use DataflowBuildConfig() and build_dataflow_cfg() functions to build a deployment package instead of a custom flow, but I get a similar error : 
```
AssertionError: DWC OutWidth is bigger than InWidth and is not divisible by it.
                Please adjust PE and SIMD values so that OutWidth % InWidth = 0
                or alternatively use impl_style = vivado
```

Do you know what is going wrong?

Julien

**Comment by julienmas** [comment]
up

**Reply by MFJI** [reply]
I also met this problem.

**Reply by shakeelakram00** [reply]
I am also having the similar problem, did you found any solution for that?


---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
/opt/finn/deps/qonnx/src/qonnx/custom_op/base.py in set_nodeattr(self, name, value)
    102         try:
--> 103             (dtype, req, def_val, allowed_values) = self.get_nodeattr_def(name)
    104             if allowed_values is not None:

/opt/finn/deps/qonnx/src/qonnx/custom_op/base.py in get_nodeattr_def(self, name)
     48         allowed_values = None
---> 49         attrdef = self.get_nodeattr_types()[name]
     50         if len(attrdef) == 3:

KeyError: 'inFIFODepth'

During handling of the above exception, another exception occurred:

AttributeError                            Traceback (most recent call last)
<ipython-input-17-937e6fa8986e> in <module>
     17     fcl_inst.set_nodeattr("PE", pe)
     18     fcl_inst.set_nodeattr("SIMD", simd)
---> 19     fcl_inst.set_nodeattr("inFIFODepth", ififodepth)
     20 
     21 # use same SIMD values for the sliding window operators

/opt/finn/deps/qonnx/src/qonnx/custom_op/base.py in set_nodeattr(self, name, value)
    137                 self.onnx_node.attribute.append(attr_proto)
    138         except KeyError:
--> 139             raise AttributeError("Op has no such attribute: " + name)
    140 
    141     def make_const_shape_op(self, shape):

AttributeError: Op has no such attribute: inFIFODepth

**Comment by anguyen9630** [comment, marked as answer]
I know this is a bit late but I finally managed to get it working so I am going to share for anyone who still needed it:
Error 1: `AssertionError: SIMD must divide IFMChannels`: If I remember correctly, this is when **the convolution input generators** IFMChannels % SIMD != 0. You must open up your ONNX file using Netron and check that.

Error 2: `AssertionError: DWC OutWidth is bigger than InWidth and is not divisible by it.`: This one I do not know how to fix the issue it self, because some of the modules you cannot set PE and some you cannot set SIMD so technically it is never achievable. **HOWEVER, the solution is to instead use the Vivado config as suggested.** Use `force_rtl_conv_inp_gen  = True` and `rtlsim_use_vivado_comps = True` in the build config and then change your **ENSURE THAT YOUR SIMD COUNT IS <= 512**
